File: source_funcs/go_transcoder/HOW_TO_CHECK_IF_A_GIVEN_ARRAY_REPRESENTS_A_BINARY_HEAP/HOW_TO_CHECK_IF_A_GIVEN_ARRAY_REPRESENTS_A_BINARY_HEAP.go
package main

func f_gold(arr []int, i int, n int) bool {
	if i > (n-2)/2 {
		return true
	}
	if arr[i] >= arr[i*2+1] && arr[i] >= arr[i*2+2] && f_gold(arr, i*2+1, n) && f_gold(arr, i*2+2, n) {
		return true
	}
	return false
}

File: rust_funcs/go_transcoder/HOW_TO_CHECK_IF_A_GIVEN_ARRAY_REPRESENTS_A_BINARY_HEAP/HOW_TO_CHECK_IF_A_GIVEN_ARRAY_REPRESENTS_A_BINARY_HEAP.rs
fn min(x: i32, y: i32) -> i32 {if x < y { x } else { y } }
fn max(x: i32, y: i32) -> i32 {if x > y { x } else { y }}
fn cmpfunc(a: &i32, b: &i32) -> std::cmp::Ordering {a.cmp(b)}
fn len(arr: &[i32]) -> usize {arr.len()}
fn sort(arr: &mut [i32]) {arr.sort_by(cmpfunc);}

use std::f32;

fn f_gold(arr: [f32;2], i: u32, n: u32) -> i32 {
    if i > (n - 2) / 2 {
        return 1;
    }
    if arr[i] >= arr[2 * i + 1] && arr[i] >= arr[2 * i + 2] && 
        f_gold(arr, 2 * i + 1, n) == 1 && f_gold(arr, 2 * i + 2, n) == 1 {
        return 1;
    }
    0
}
